bindings/cs/MongoDB.Crypt/PinnedBinary.cs

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright 2019–present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System;
+
+namespace MongoDB.Crypt
+{
+    /// <summary>
+    /// A handle to a binary that must be either in unsafe code or pinned by the GC.
+    /// </summary>
+    /// <seealso cref="MongoDB.Crypt.Binary" />
+    internal class PinnedBinary : Binary
+    {
+        internal PinnedBinary(IntPtr ptr, uint len)
+            : base(Library.mongocrypt_binary_new_from_data(ptr, len))
```

Comment:
I did not have a good design for this particular pattern so I named the class PinnedBinary in the hope it would not be misused.

When the .net passes byte buffers into libmongocrypt, it uses mongocrypt_binary_t which contains a pointer and length. ANY clr memory passed to libmongocrypt has to be pinned so that the GC does not move it so this is my way of conveying it. 

Now, libmongocrypt does not hold onto the memory after the call, it just needs it pinned during the duration of the call.

This same class is used for receiving memory from libmongocrypt but that is less tricky to use since the GC does not interfere with malloc

Reply:
@vincent if that explanation is sufficient for you perhaps you can resolve.

I'm still confused (not by the concept of pinning, but by this class being a subclass of Binary).

Reply:
So basically this class seems to have no reason to exist. It doesn't pin memory, it doesn't verify that the memory passed to the constructor is pinned or not, and it doesn't unpin memory.

It seems to be used solely as a signal to the reader somewhat to the effect of "this memory is probably pinned but I'm not helping you in any way or even checking whether it is or not.".

Since this class seems to effectively do nothing I have no objection to leaving it as is given time constraints.